### PR TITLE
Start and Stop methods for gateway moved to node methods

### PR DESCRIPTION
**Pull request by carsonfarmer**
This should allow us to gracefully stop the gateway server before shutting down

### CODE REVIEW

**Comment by carsonfarmer** [comment]
huh, not quite sure how to interpret the failing tests on this @sanderpick? should be gtg otherwise?

**Comment by sanderpick** [comment]
you should be able to run the mobile tests by just clicking those green arrows (the very top one in github.com/textileio/textile-go/mobile_test and github.com/textileio/textile-go/core_test)... failing cause mobile does not have the start gateway method anymore

File: core/core.go [hunk 1]
```diff
@@ -90,17 +90,25 @@ type TextileNode struct {
...
+	// Gateway http.Server
+
+	// The local decrypting gateway server
+	GatewayProxy *http.Server
+
+	// The local password used to authenticate secure GatewayProxy requests
+	GatewayPassword string
 }
 
+// PhotoList is a JSON-type structure that contains a list of photo hashes
```
[the remark this answers is not in the document]
Comment (carsonfarmer) [reply core/core.go]: haha, my ide was complaining

File: core/core.go [hunk 2]
```diff
@@ -58,10 +60,15 @@ var Node *TextileNode
 
 const roomRepublishInterval = time.Minute
 const pingRelayInterval = time.Second * 30
-const kPingTimeout = 10 * time.Second
+const pingTimeout = 10 * time.Second
```
Comment (sanderpick) [inline core/core.go]: nice, i don't remember why the "k" was there ha

File: core/core.go [hunk 3]
```diff
@@ -58,10 +60,15 @@ var Node *TextileNode
...
 
+// ErrNodeRunning is an error for when node start is called on a running node
 var ErrNodeRunning = errors.New("node is already running")
 
+// ErrNodeNotRunning is an error for  when node stop is called on a nil node
+var ErrNodeNotRunning = errors.New("node is not running")
```
Comment (sanderpick) [inline core/core.go]: sweet

File: core/core.go [hunk 4]
```diff
@@ -200,6 +220,55 @@ func NewNode(repoPath string, isMobile bool, logLevel logging.Level) (*TextileNo
 	return node, nil
 }
 
+// ServeHTTPGatewayProxy starts the secure HTTP gatway proxy server
+func startGatewayProxy(t *TextileNode) (<-chan error, error) {
```
Comment (sanderpick) [inline core/core.go]: nice

File: core/core.go [hunk 5]
```diff
@@ -235,6 +304,29 @@ func (t *TextileNode) Start() error {
...
+	}
+
+	go func() {
+		for {
+			select {
+			case err, ok := <-gwpErrc:
+				if err != nil {
+					log.Errorf("service error: %s", err)
+				}
+				if !ok {
```
Comment (sanderpick) [inline core/core.go]: boom

File: mobile/mobile.go [hunk 6]
```diff
@@ -59,17 +59,6 @@ func (w *Wrapper) Start() error {
...
-	}
-	if _, err := tcore.ServeHTTPGatewayProxy(w.node); err != nil {
-		return err
-	}
-	w.gatewayRunning = true
-	return nil
-}
-
 func (w *Wrapper) Stop() error {
 	return w.node.Stop()
```
Comment (sanderpick) [inline mobile/mobile.go]: we'll need to catch the `ErrNodeNotRunning` error here, right? i can do in my branch though

Comment (sanderpick) [reply mobile/mobile.go]: (since we don't want to error out in RN)

**Pull request merged**